app/exec/build/tasks/sign.ts

```diff
@@ -0,0 +1,159 @@
+		const taskZipPath = taskZipsPath[0];
+		const certFingerprint: string | null = await this.commandArgs.certFingerprint.val();
+		const newGuid: string | null = await this.commandArgs.newGuid.val();
+		const newNameSuffix: string | null = await this.commandArgs.newNameSuffix.val();
+
+		// verify that we can find NuGet
+		const nuGetPath: string = shell.which('nuget');
+		if (!nuGetPath) {
+			throw new Error('Unable to find NuGet. Please add NuGet to the PATH before continuing.');
+		}
+
+		const tempFolder: string = fs.mkdtempSync(path.join(os.tmpdir(), 'task-'));
+		let taskTempFolder: string = path.join(tempFolder, 'task');
+		const taskTempZipPath: string = path.join(tempFolder, 'task.zip');
+		const taskTempNupkgPath: string = path.join(tempFolder, 'task.nupkg');
+
+		// Create temp folder
+		if (fs.existsSync(tempFolder)) {
+			await del(tempFolder, { force: true });
+		}
+
+		fs.mkdirSync(tempFolder);
+		fs.mkdirSync(taskTempFolder);
+
+		// Copy task contents to temp folder
+		shell.cp('-R', taskZipPath, taskTempFolder);
+
+		// Task temp folder is now task\PREVIOUS_NAME
+		taskTempFolder = path.join(taskTempFolder, path.basename(taskZipPath));
+
+		// Update guid/name
+		const needToUpdateTaskJson: boolean = !!newGuid || !!newNameSuffix;
+		if (needToUpdateTaskJson) {
+			const taskJsonPath: string = path.join(taskTempFolder, 'task.json');
+			const data: string = fs.readFileSync(taskJsonPath, 'utf8');
+			let taskJson: any = JSON.parse(data);
+
+			if (newGuid) {
+				taskJson.id = newGuid;
+			}
+
+			if (newNameSuffix) {
+				taskJson.name = `${taskJson.name}${newNameSuffix.replace(/'/g, "")}`;
+			}
+
+			fs.writeFileSync(taskJsonPath, JSON.stringify(taskJson, null, 4));
+		}
+
+		// Zip
+		await this.zipDirectory(taskTempFolder, taskTempZipPath);
+
+		// Rename to nupkg
+		fs.renameSync(taskTempZipPath, taskTempNupkgPath);
+		
+		// Sign
+		const command: string = `"${nuGetPath}" sign ${taskTempNupkgPath} -CertificateFingerprint ${certFingerprint} -NonInteractive`;
+		trace.info(`Executing command: '${command}'`);
+		const execResult: any = shell.exec(command, { silent: true });
+		if (execResult.code === 1) { 
+			trace.info(execResult.output);
+			await del(tempFolder, { force: true });
+		
+			const result: TaskSignResult = { signingSuccessful: false };
+			return result;
+		}
+
+		// Rename .nupkg back to to .zip
+		fs.renameSync(taskTempNupkgPath, taskTempZipPath);
+
+		// Extract into new temp task folder
+		const taskAfterSignTempFolder: string = path.join(tempFolder, 'task-after-sign');
+		fs.mkdirSync(taskAfterSignTempFolder);
+		var zip = new admZip(taskTempZipPath);
+		zip.extractAllTo(taskAfterSignTempFolder);
+
+		// Copy task contents
+		// This can include the new signature file as well as a modified task.json
+		await this.ncpAsync(taskAfterSignTempFolder, taskZipPath);
+
+		// Delete temp folder
+		await del(tempFolder, { force: true });
```

Comment:
Rather than doing this in a couple places, maybe it makes sense to wrap this whole function in a try...finally block that does this? That also will help in error cases